lucene/core/src/java/org/apache/lucene/util/GroupVIntUtil.java

```diff
@@ -56,7 +58,54 @@ public static void readGroupVInts(DataInput in, int[] dst, int limit) throws IOE
    * @param dst the array to read ints into.
    * @param offset the offset in the array to start storing ints.
    */
-  public static void readGroupVInt(DataInput in, int[] dst, int offset) throws IOException {
+  private static void readGroupVInt(DataInput in, int[] dst, int offset) throws IOException {
+    final int flag = in.readByte() & 0xFF;
+
+    final int n1Minus1 = flag >> 6;
+    final int n2Minus1 = (flag >> 4) & 0x03;
+    final int n3Minus1 = (flag >> 2) & 0x03;
+    final int n4Minus1 = flag & 0x03;
+
+    // if our DataInput implements RandomAccessInput for absolute access and IndexInput for seeking,
+    // we use a branch-less implementation:
+    if (in instanceof RandomAccessInput rin && in instanceof IndexInput iin) {
+      long pos = iin.getFilePointer();
+      if (iin.length() - pos >= MAX_LENGTH_PER_GROUP - 1) {
```

Comment:
FWIW the following reads a bit easier to me. It's very clear since it reads 4 integers a few lines below. Versus understanding what this constant is and why we need to subtract one.

```suggestion
      if (iin.length() - pos >= 4 * Integer.BYTES) {
```